CIS crashes with "panic: reflect: slice index out of range" after overlapping OS routes are supposed by CIS

#### Setup Details
 CIS Version : 2.12.1      
 Build: f5networks/k8s-bigip-ctlr:2.12.1  
 BIGIP Version: Big IP 15.1.8.1.0.32.3    
AS3 Version: 3.34
Agent Mode: AS3
Orchestration: K8S/OSCP v4.9.xx, Kubernetes v1.24.0+3882f8f
Orchestration Version:
Pool Mode: Cluster

#### Description  
I use the CIS in legacy mode (OpenShift Routes).
The CIS crashes after overlapping OS routes is supposed by CIS, but in reality there are no overlapping OS routes!

#### Steps To Reproduce
1) Add new OS-route-01
2) Delete this new OS-route-01
3) Add new OS-route-02
4) Wait for CIS crash (may take hours)

Time between step 2 + 3 = couple of minutes (max. 5mins)

OS-Route-01: admin-ui-xxx-internal-internal
OS-Route-02: admin-ui-xxx-internal-external
Both OS-Routes use same URI xxxde/adminui/xxx-internal


#### Expected Result 
CIS should detect the deletion of OS-route-01 in step 2 in a timely and correct manner. It should accept OS-route-02 in step 3 and should continue operating normally.


#### Actual Result  
After step 3, error messages are displayed erroneously. CIS crashes with panic-error after some time.

#### Diagnostic Information
```buildoutcfg
2023/05/31 03:43:25 [ERROR] Discarding route admin-ui-xxx-internal-external as other route already exposes URI xxxde/adminui/xxx-internal and is older 
2023/05/31 03:43:26 [ERROR] Discarding route admin-ui-xxx-internal-external as other route already exposes URI xxxde/adminui/xxx-internal and is older 
2023/05/31 03:43:26 [ERROR] Discarding route admin-ui-xxx-internal-external as other route already exposes URI xxxde/adminui/xxx-internal and is older 

panic: reflect: slice index out of range [recovered]
	panic: reflect: slice index out of range

goroutine 17069321 [running]:
encoding/json.(*encodeState).marshal.func1()
	/usr/local/go/src/encoding/json/encode.go:327 +0x6e
panic({0x15d2560, 0x1ae03c0})
	/usr/local/go/src/runtime/panic.go:884 +0x212
reflect.Value.Index({0x1589620?, 0xc000b7e528?, 0xc0003e06e0?}, 0x15?)
	/usr/local/go/src/reflect/value.go:1412 +0x16d
encoding/json.arrayEncoder.encode({0x15ce120?}, 0xc000b88000, {0x1589620?, 0xc000b7e528?, 0xc003d120c0?}, {0x80?, 0xcd?})
	/usr/local/go/src/encoding/json/encode.go:915 +0xae
encoding/json.sliceEncoder.encode({0x15ce1e0?}, 0xc000b88000, {0x1589620?, 0xc000b7e528?, 0xc000b7e4c8?}, {0xa?, 0x0?})
	/usr/local/go/src/encoding/json/encode.go:888 +0x367
encoding/json.structEncoder.encode({{{0xc0026b9320?, 0x45379c?, 0x15d2560?}, 0xc00353b380?}}, 0xc000b88000, {0x16e5100?, 0xc000b7e528?, 0x9?}, {0x0, 0x1})
	/usr/local/go/src/encoding/json/encode.go:760 +0x1f4
encoding/json.structEncoder.encode({{{0xc004415200?, 0x404f54?, 0x26eaac0?}, 0xc00353b3e0?}}, 0xc000b88000, {0x1779080?, 0xc000b7e380?, 0x189acbb?}, {0x0, 0x1})
	/usr/local/go/src/encoding/json/encode.go:760 +0x1f4
encoding/json.ptrEncoder.encode({0xc000b88000?}, 0xc000b88000, {0x18622e0?, 0xc000b7e380?, 0x18622e0?}, {0x86?, 0xbb?})
	/usr/local/go/src/encoding/json/encode.go:944 +0x25e
encoding/json.(*encodeState).reflectValue(0x0?, {0x18622e0?, 0xc000b7e380?, 0x40f467?}, {0x78?, 0x0?})
	/usr/local/go/src/encoding/json/encode.go:359 +0x78
encoding/json.(*encodeState).marshal(0x2?, {0x18622e0?, 0xc000b7e380?}, {0x88?, 0x38?})
	/usr/local/go/src/encoding/json/encode.go:331 +0xfa
encoding/json.(*Encoder).Encode(0xc0013c18c0, {0x18622e0, 0xc000b7e380})
	/usr/local/go/src/encoding/json/stream.go:206 +0x7e
k8s.io/apimachinery/pkg/runtime/serializer/json.(*Serializer).doEncode(0x188c17a?, {0x1ae6928?, 0xc000b7e380?}, {0x1ae1d60, 0xc0056ea270})
	/go/src/github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go:327 +0x1b5
k8s.io/apimachinery/pkg/runtime/serializer/json.(*Serializer).Encode(0xc0000b5d60, {0x1ae6928, 0xc000b7e380}, {0x1ae1d60, 0xc0056ea270})
	/go/src/github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go:301 +0xfc
k8s.io/apimachinery/pkg/runtime.WithVersionEncoder.Encode({{0x1aeac30, 0xc0004de180}, {0x1ae9970, 0xc0000b5d60}, {0x1ae98d0, 0xc00028f960}}, {0x1ae6928, 0xc000b7e380}, {0x1ae1d60, 0xc0056ea270})
	/go/src/github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/apimachinery/pkg/runtime/helper.go:240 +0x2ee
k8s.io/apimachinery/pkg/runtime.Encode({0x1aeabb8, 0xc0056ea240}, {0x1ae6928, 0xc000b7e380})
	/go/src/github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/apimachinery/pkg/runtime/codec.go:50 +0x64
k8s.io/client-go/rest.(*Request).Body(0xc0002900f0, {0x18622e0?, 0xc000b7e380?})
	/go/src/github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/rest/request.go:453 +0x48b
github.com/openshift/client-go/route/clientset/versioned/typed/route/v1.(*routes).UpdateStatus(0xc007f94e60, {0x1af7e28, 0xc000120008}, 0xc000b7e380, {{{0x0, 0x0}, {0x0, 0x0}}, {0x0, 0x0, ...}, ...})
	/go/src/github.com/F5Networks/k8s-bigip-ctlr/vendor/github.com/openshift/client-go/route/clientset/versioned/typed/route/v1/route.go:133 +0x1a7
github.com/F5Networks/k8s-bigip-ctlr/v2/pkg/appmanager.(*Manager).updateRouteAdmitStatus(0xc00023a000, {0xc00717ec80, 0x32}, {0x189a001, 0x12}, {0xc00427f080, 0xa6}, {0x188cc1a, 0x5})
	/go/src/github.com/F5Networks/k8s-bigip-ctlr/pkg/appmanager/agentResponseHandler.go:153 +0x490
created by github.com/F5Networks/k8s-bigip-ctlr/v2/pkg/appmanager.(*Manager).syncRoutes
	/go/src/github.com/F5Networks/k8s-bigip-ctlr/pkg/appmanager/appManager.go:1990 +0x9ea
```  

#### Observations (if any)

--- a/docs/RELEASE-NOTES.rst
+++ b/docs/RELEASE-NOTES.rst
@@ -17,6 +17,9 @@ Bug Fixes
 * `Issue 2867 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/2867>`_: Ignore virtualServerName if hostGroup configured
 * `Issue 2898 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/2898>`_: Fix for CIS crash with namespace-label parameter
 * `Issue 2778 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/2778>`_: Fix hostless VS with IPAM to work with hostgroup
+* `Issue 2908 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/2908>`_: Fix for CIS crash while updating the route status
+
+
 2.13.0
 -------------
 

--- a/pkg/appmanager/agentResponseHandler.go
+++ b/pkg/appmanager/agentResponseHandler.go
@@ -89,7 +89,7 @@ func (appMgr *Manager) updateRouteAdmitStatusAll() {
 		return
 	}
 	// Check whether we are processing this route.
-	// Else, clean the route metadata if we add any in past.
+	// Else, clean the route metadata if we add any in the past.
 	for _, aRoute := range allRoutes.Items {
 		routeKey := fmt.Sprintf("%v/%v", aRoute.Namespace, aRoute.Name)
 		if _, ok := processedRoutes[routeKey]; !ok {
@@ -117,6 +117,11 @@ func (appMgr *Manager) updateRouteAdmitStatus(
 	message string,
 	status v1.ConditionStatus,
 ) {
+	defer func() {
+		if r := recover(); r != nil {
+			log.Errorf("CIS recovered from the panic caused by route status update: %v\n")
+		}
+	}()
 	for retryCount := 0; retryCount < 3; retryCount++ {
 		route := appMgr.fetchRoute(rscKey)
 		if route == nil {

--- a/pkg/appmanager/appManager.go
+++ b/pkg/appmanager/appManager.go
@@ -730,6 +730,7 @@ type serviceQueueKey struct {
 	ResourceKind string
 	ResourceName string
 	Operation    string
+	Object       interface{}
 }
 
 type nodeInformer struct {
@@ -1073,11 +1074,11 @@ func (appMgr *Manager) enqueueIngressUpdate(cur, old interface{}, operation stri
 
 func (appMgr *Manager) enqueueRoute(obj interface{}, operation string) {
 	if ok, keys := appMgr.checkValidRoute(obj); ok {
-		if operation == OprTypeDelete {
-			appMgr.deleteHostPathMapEntry(obj)
-		}
 		for _, key := range keys {
 			key.Operation = operation
+			if operation == OprTypeDelete {
+				key.Object = obj
+			}
 			appMgr.vsQueue.Add(*key)
 		}
 	}
@@ -2023,6 +2024,9 @@ func (appMgr *Manager) syncRoutes(
 	appInf *appInformer,
 	dgMap InternalDataGroupMap,
 ) error {
+	if sKey.Operation == OprTypeDelete && sKey.ResourceKind == Routes {
+		appMgr.deleteHostPathMapEntry(sKey.Object)
+	}
 	routeByIndex, err := appInf.getOrderedRoutes(sKey.Namespace)
 	if nil != err {
 		log.Warningf("[CORE] Unable to list routes for namespace '%v': %v",
@@ -3613,18 +3617,17 @@ func (appMgr *Manager) updateHostPathMap(timestamp metav1.Time, key string) {
 func (appMgr *Manager) deleteHostPathMapEntry(obj interface{}) {
 	// This function deletes the route entry from processedHostPath
 	route := obj.(*routeapi.Route)
+	var key string
+	if route.Spec.Path == "/" || len(route.Spec.Path) == 0 {
+		key = route.Spec.Host + "/"
+	} else {
+		key = route.Spec.Host + route.Spec.Path
+	}
 	appMgr.processedHostPath.Lock()
 	defer appMgr.processedHostPath.Unlock()
-	for hostPath, routeTimestamp := range appMgr.processedHostPath.processedHostPathMap {
-		var key string
-		if route.Spec.Path == "/" || len(route.Spec.Path) == 0 {
-			key = route.Spec.Host + "/"
-		} else {
-			key = route.Spec.Host + route.Spec.Path
-		}
-		if routeTimestamp == route.CreationTimestamp && hostPath == key {
-			// Deleting the ProcessedHostPath map if route's path is changed
-			delete(appMgr.processedHostPath.processedHostPathMap, hostPath)
+	if routeTimestamp, ok := appMgr.processedHostPath.processedHostPathMap[key]; ok {
+		if routeTimestamp == route.CreationTimestamp {
+			delete(appMgr.processedHostPath.processedHostPathMap, key)
 		}
 	}
 }

--- a/pkg/appmanager/appManager_test.go
+++ b/pkg/appmanager/appManager_test.go
@@ -2672,7 +2672,7 @@ var _ = Describe("AppManager Tests", func() {
 				mockMgr.appMgr.updateHostPathMap(route1.CreationTimestamp, key)
 				_, ok := mockMgr.appMgr.processedHostPath.processedHostPathMap[key]
 				Expect(ok).To(BeTrue())
-				mockMgr.appMgr.enqueueRoute(route1, OprTypeDelete)
+				mockMgr.appMgr.deleteHostPathMapEntry(route1)
 				_, ok = mockMgr.appMgr.processedHostPath.processedHostPathMap[key]
 				Expect(ok).To(BeFalse())
 				spec2 := routeapi.RouteSpec{

--- a/pkg/controller/nativeResourceWorker.go
+++ b/pkg/controller/nativeResourceWorker.go
@@ -1463,6 +1463,11 @@ func (ctlr *Controller) updateRouteAdmitStatus(
 	message string,
 	status v1.ConditionStatus,
 ) {
+	defer func() {
+		if r := recover(); r != nil {
+			log.Errorf("CIS recovered from the panic caused by route status update: %v\n")
+		}
+	}()
 	for retryCount := 0; retryCount < 3; retryCount++ {
 		route := ctlr.fetchRoute(rscKey)
 		if route == nil {

